addons/binding/org.openhab.binding.chamberlainmyq/src/main/java/org/openhab/binding/chamberlainmyq/handler/ChamberlainMyQHandler.java

```diff
@@ -0,0 +1,157 @@
+     *
+     * @return this thing gateway handler, null if it hasn't been set yet.
+     */
+    protected ChamberlainMyQGatewayHandler getGatewayHandler() {
+        Bridge gateway = getBridge();
+        return gateway == null ? null : (ChamberlainMyQGatewayHandler) gateway.getHandler();
+    }
+
+    /**
+     * Function called when a communication error between the gateway and the thing has been detected.
+     */
+    protected void handleCommunicationError(String errorMessage) {
+        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, errorMessage);
+    }
+
+    private abstract class ChamberlainMyQDeviceRequestCallback implements RequestCallback {
+        /**
+         * Handler of the thing that should for which the configuration should be set.
+         */
+        protected ChamberlainMyQHandler handler;
+
+        /**
+         * Creates a new instance of this class.
+         *
+         * @param handler The handler for which the configuration should be read.
+         */
+        public ChamberlainMyQDeviceRequestCallback(ChamberlainMyQHandler handler) {
+            if (handler != null) {
+                this.handler = handler;
+            }
+        }
+
+        @Override
+        public void onError(String error) {
+            handler.handleCommunicationError(error);
+        }
+    }
+
+    //////////// Read State functions ////////////
+
+    /**
+     * Specialization of a {link RequestCallback} to read a device configuration.
+     *
+     * @author Scott Hanson
+     */
+    private class ReadDeviceStateCallback extends ChamberlainMyQDeviceRequestCallback {
+        /**
+         * Creates a new instance of this class.
+         *
+         * @param handler The handler for which the state should be read.
+         */
+        public ReadDeviceStateCallback(ChamberlainMyQHandler handler) {
+            super(handler);
+        }
+
+        @Override
+        public void parseRequestResult(MyqJson jsonResult) {
+            logger.trace("Parsing a ReadDeviceState request result: {}", jsonResult);
+            // The response from the server is a JSON object containing the device information and state.
+            handler.updateDeviceStateCallback(jsonResult);
+        }
+    }
+
+    /**
+     * Query the {@link MyQGatewayHandler} for this device's state.
+     */
+    protected void readDeviceState() {
+        try {
+            getGatewayHandler().sendRequestToServer(new ReadDeviceStateCallback(this));
+        } catch (IOException e) {
+            logger.warn("Error while querying the myQ Service for {}", e);
+        }
+    }
+
+    /**
+     * Callback called once the device state has been updated.
+     *
+     * @param jsonDataBlob the reply from the gateway.
+     */
+    protected void updateDeviceStateCallback(MyqJson myqDevices) {
+        String deviceID = getThing().getProperties().get(MYQ_ID);
+
+        for (Device myqdevice : myqDevices.getDevices()) {
```

Comment:
Can you describe how the binding does its internal bridge/thing communication, because it feels different